pkg/component/runtime/manager_fake_input_test.go

```diff
@@ -3249,131 +3131,76 @@ func (suite *FakeInputSuite) TestManager_OutputChange() {
+		"the 1st component model")
+
+	// Wait the 1st config to be applied and the comp0ID to start
+	require.Eventuallyf(t,
+		func() bool {
+			filtered := logs.FilterMessageSnippet(
+				fmt.Sprintf("Starting component %q", comp0ID)).
+				TakeAll()
+			return len(filtered) > 0
+		},
+		30*time.Second,
+		200*time.Millisecond,
+		"component %s did not start", comp0ID)
 
-	updateSleep := 300 * time.Millisecond
-	if runtime.GOOS == component.Windows {
-		// windows is slow, preventing flakiness
-		updateSleep = time.Second
-	}
-	time.Sleep(updateSleep)
 	m.Update(component.Model{Components: components2})
 	err = <-m.errCh
-	require.NoError(t, err)
+	require.NoError(t, err, "expected no error from the manager when applying"+
+		"the 2nd component model")
+
+	// Wait the 2nd config to be applied and the comp1ID to start
+	require.Eventuallyf(t,
+		func() bool {
+			filtered := logs.FilterMessageSnippet(
+				fmt.Sprintf("Starting component %q", comp1ID)).
+				TakeAll()
+			return len(filtered) > 0
+		},
+		30*time.Second,
+		200*time.Millisecond,
+		"component %s did not start", comp1ID)
 
-	count := 0
-	timeout := 30 * time.Second
-	endTimer := time.NewTimer(timeout)
-	defer endTimer.Stop()
+	// component 1 started, we can stop the manager
+	cancel()
 
-LOOP:
-	for {
-		select {
-		case <-endTimer.C:
-			t.Fatalf("timed out after %s seconds, "+
-				"did not receive enought state changes", timeout)
-		case err := <-errCh:
-			require.NoError(t, err)
-		case err := <-subErrCh0:
-			t.Logf("[subErrCh0] received: %v", err)
-			require.NoError(t, err)
-			count++
-			if count >= 2 {
-				break LOOP
-			}
-		case err := <-subErrCh1:
-			t.Logf("[subErrCh1] received: %v", err)
-			require.NoError(t, err)
-			count++
-			if count >= 2 {
-				break LOOP
-			}
-		}
-	}
+	comp0StartLogs := logs.FilterMessageSnippet(
+		fmt.Sprintf("Starting component %q", comp0ID)).TakeAll()
+	comp0StopLogs := logs.FilterMessageSnippet(
+		fmt.Sprintf("Stopping component %q", comp0ID)).TakeAll()
+	comp1StartLogs := logs.FilterMessageSnippet(
+		fmt.Sprintf("Starting component %q", comp1ID)).TakeAll()
 
-	subCancel()
-	cancel()
+	assert.Len(t, comp0StartLogs, 1,
+		"component %d started more than once", comp0ID)
+	assert.Len(t, comp0StopLogs, 1,
+		"component %d stopped more than once", comp0ID)
+	assert.Len(t, comp1StartLogs, 1,
+		"component %d started more than once", comp1ID)
 
-	// check progression, require stop fake-0 before start fake-1
-	stateProgressionWG.Wait()
-	comp0Stopped := false
-	for _, step := range stateProgression {
-		if step.componentID == IDComp0 &&
-			step.state.State == client.UnitStateStopped {
-			comp0Stopped = true
-		}
-		if step.componentID == IDComp1 &&
-			step.state.State == client.UnitStateStarting {
-			require.True(t, comp0Stopped)
-			break
+	assert.Truef(t, comp0StopLogs[0].Time.Before(comp1StartLogs[0].Time),
+		"component %s stopped after %s", comp0ID, comp1ID)
```

Comment:
I'm not a fan of relying on logs for tests, but due to the concurrent nature of how the state is reported and the test was collecting them, checking directly the logs are the simplest and mo reliable way to check the correct order of the events.

Reply:
Agree this is fine.

You could additionally verify that the component 0 start time is before the component 0 stop time.

Right now the following ordering is allowed by the test (even though our logic should not produce it):

comp0stop -> comp1start -> comp0start